### DISCUSSION TITLE
Can create function accept a class?

### DISCUSSION THREAD

**Discussion by alfredsgenkins** [pmndrs/zustand#1323] [Ideas]
Hey!

From my observations, it seems like a `create` function can easily accept a function-constructing-a-class argument. Maybe can pass a class directly, have not really tried, more a theoretical question at 5AM... So:

```js
import create from 'zustand';

class BearStore {
  constructor(set) {
    this.set = set;
  }

  bears = 0;

  increasePopulation: () => this.set(this.increasePopulation)

  _increasePopulation: (state) => ({ bears: state.bears + 1 }))

  removeAllBears: () => this.set({ bears: 0 })
}

const useBearStore = create((set) => new BearStore(set))
```

**Looking at this monster, you might think: Why to accept classes?**

Well, if we assume store functions could need modification in the future, that would work really well. As we could extend them easily.

**Composition is possible...**

Yeah, we could try composition too, but the boilerplate + a ton of set calls, will it be worth it? Like really, we want to expose each setter for extension or interception...

For context: the apps we build are not single-use. We build a base for hundreds other websites (ecom specifically). Developers build on-top of main codebase, without modifying the source code. There are direct overrides + extensions and plugin approach (intercepting any function).

So given I need extensibility, can I use Zustand with class as creator? There are quite few optimisations possible, we can always bind set to selected functions, and more... Will this break performance? Any consequences of doing this?

**Comment by dai-shi** [comment]
somebody raised the similar idea. 

https://github.com/pmndrs/zustand/discussions/750#discussioncomment-3676120
